[scripts-audit] CMake buildsystem!!!

strega-nil opened the pull request:
See PR #16055, issue #16188

Depends on PR #16419

Part of the "merge 2021-02-26 EOD" group.

cc @ras0219, @Neumann-A for review

Neumann-A commented on ports/fmt/CONTROL:
```diff
@@ -1,5 +1,6 @@
+Port-Version: 1
 Homepage: https://github.com/fmtlib/fmt
 Description: Formatting library for C++. It can be used as a safe alternative to printf or as a fast alternative to IOStreams.
+Build-Depends: vcpkg-cmake-buildsystem
```
isn't `vcpkg-cmake` enough? (or use `vcpkg-scripts-cmake`).
I don't feel like `buildsystem` is the right descriptor here.

strega-nil replied:
I'm not sure, I just kinda chose a name because I needed to; please post your thoughts on a better name.

Neumann-A commented on ports/fmt/portfile.cmake:
```diff
@@ -1,3 +1,5 @@
+include(vcpkg-cmake-buildsystem)
```
please avoid this somehow. (e.g. by passing the dependency?)

strega-nil replied:
I don't really want to avoid this, and I honestly can't think of how?

Neumann-A replied:
The `Build-Depends:` already has the information for the buildsystem. vcpkg needs to pass the information (+feature info of depends) to cmake any way so that some downstream ports might react to the linkage difference of different depends/features. If you have the dep info in the cmake script available you can just hide the `include` here away. 

also if you only use `include(vcpkg-cmake-buildsystem)` here users are able to override `CMAKE_MODULE_PATH` in a triplet.

Neumann-A commented on ports/fmt/portfile.cmake:
```diff
@@ -6,16 +8,17 @@ vcpkg_from_github(
     PATCHES fix-warning4189.patch
 )
-vcpkg_configure_cmake(
+vcpkg_cmake_buildsystem_configure(
```
If you need another name why not `vcpkg_cmake_configure`. 
Also scripts need to be versioned so having `vcpkg_configure_cmake` use `cmake_language(CALL <command> [<arg>...])` and have command be something like `vcpkg_configure_cmake_<version>`. The reason behind this is that probably multiple versions of the scripts are required.

Neumann-A replied:
scripts somehow does not work well with the single version behavior vcpkg currently has if you want to support multiple versions of ports which might depend on different script versions.

strega-nil replied:
This does not make any sense to me.

Neumann-A replied:
port1#V1 depends on scripts#V4; port2#V15 depends on scripts#V2; 
port1#V1 does not build with scripts#V2 and port2#V15 does not build with scripts#V4 (because somebody broke backcompat) for some reason (port2#V50 does but that does not matter here.)

scripts are a bit different to normal port dependencies, in my opinion. You can have multiple version of scripts at the same time if you version the folders|files|functions without bigger problems.

Neumann-A commented on ports/fmt/portfile.cmake:
```diff
@@ -40,7 +43,7 @@ if(VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
 file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
 
-vcpkg_fixup_cmake_targets()
+vcpkg_cmake_buildsystem_fixup_targets()
```
Although targets are predominantly generated by cmake there is a high possibility that meson will also generate configs/targets. I also saw autoconf generate a config for something. So the name here is wrong.

strega-nil replied:
that's fair enough.

strega-nil committed:
[scripts-audit] Documentation Generation Changes

* pull the cmake doc comment parsing out into its own function
* support cmake helper ports
* add real support for deprecation, as opposed to ad-hoc

strega-nil committed:
[scripts-audit] CMake buildsystem!!!

Deprecate `vcpkg_*_cmake` in favor of `vcpkg_cmake_*` from the
`vcpkg-cmake` port, as well as `vcpkg_fixup_cmake_targets`
in favor of `vcpkg_cmake_config_fixup` from the
`vcpkg-cmake-config` port.

strega-nil committed:
format manifest and switch to date versions

strega-nil commented:
Closing in favor of a single PR that contains all of these PRs. (to avoid destroying the CI system)

JackBoosY commented on ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:
```diff
@@ -0,0 +1,391 @@
+        set(ninja_can_be_used OFF)
+        set(ninja_host OFF)
+    elseif(targetting_uwp)
+        # Ninja and MSBuild have many differences when targetting UWP, so use MSBuild to maximize existing compatibility
+        set(ninja_can_be_used OFF)
```
That breaks the behavor, see https://github.com/microsoft/vcpkg/issues/19818
